hashes/src/chacha20poly1305/mod.rs

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Combine the ChaCha20 stream cipher with the Poly1305 message authentication code
+//! to form an authenticated encryption with additional data (AEAD) algorithm.
+
+pub mod chacha20;
+pub mod poly1305;
+
+use chacha20::ChaCha20;
+use poly1305::Poly1305;
+
+use core::fmt;
+
+/// Zero array for padding slices.
+const ZEROES: [u8; 16] = [0u8; 16];
+
+/// Errors encrypting and decrypting messages with ChaCha20 and Poly1305 authentication tags.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Error {
+    /// Additional data showing up when it is not expected.
+    UnauthenticatedAdditionalData,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::UnauthenticatedAdditionalData => write!(f, "Unauthenticated aad."),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Error::UnauthenticatedAdditionalData => None,
+        }
+    }
+}
+
+/// Encrypt and decrypt content along with a authentication tag.
+#[derive(Debug)]
+pub struct ChaCha20Poly1305 {
+    key: [u8; 32],
+    nonce: [u8; 12],
+}
+
+impl ChaCha20Poly1305 {
+    /// Make a new instance of a ChaCha20Poly1305 AEAD.
+    pub fn new(key: [u8; 32], nonce: [u8; 12]) -> Self {
+        ChaCha20Poly1305 { key, nonce }
+    }
+
+    /// Encrypt content in place and return the poly1305 16-byte authentication tag.
+    ///
+    /// # Arguments
+    ///
+    /// - `content` - Plaintext to be encrypted in place.
+    /// - `aad`     - Optional metadata covered by the authentication tag.
+    ///
+    /// # Returns
+    ///
+    /// The 16-byte authentication tag.
+    pub fn encrypt(self, content: &mut [u8], aad: Option<&[u8]>) -> [u8; 16] {
+        let mut chacha = ChaCha20::new_from_block(self.key, self.nonce, 1);
+        chacha.apply_keystream(content);
+        let keystream = chacha.get_keystream(0);
+        let mut poly = Poly1305::new(keystream[..32].try_into().expect("infallible conversion"));
+        let aad = aad.unwrap_or(&[]);
+        // AAD and ciphertext are padded if not 16-byte aligned.
+        poly.add(aad);
+        let aad_overflow = aad.len() % 16;
+        if aad_overflow > 0 {
+            poly.add(&ZEROES[0..(16 - aad_overflow)]);
+        }
+
+        poly.add(content);
+        let text_overflow = content.len() % 16;
+        if text_overflow > 0 {
+            poly.add(&ZEROES[0..(16 - text_overflow)]);
+        }
+
+        let aad_len = aad.len().to_le_bytes();
+        let msg_len = content.len().to_le_bytes();
+        let mut len_buffer = [0u8; 16];
+        len_buffer[..aad_len.len()].copy_from_slice(&aad_len[..]);
+        for i in 0..msg_len.len() {
+            len_buffer[i + aad_len.len()] = msg_len[i]
```

Comment:
Please no indexing in Rust unless absolutely required. We already had enough bugs caused by it.

Reply:
That feels a little strong, for a low-level hashing crate anyway, but agreed that in this case we could use `split_mut` to split up `len_buffer` and then `copy_from_slice` to copy the bytes. And we could use `debug_assert_eq` to check that all relevant slice lengths are consistent.

Reply:
Hm, that makes sense to me. I think I have quite a few spots to clean up.